servicenow/check.py

```diff
@@ -0,0 +1,208 @@
+            raise Exception('ServiceNow CMDB topology instance missing "basic_auth.user" value.')
+        if 'password' not in basic_auth:
+            raise Exception('ServiceNow CMDB topology instance missing "basic_auth.password" value.')
+
+        relation_types = {}
+
+        basic_auth_user = basic_auth['user']
+        basic_auth_password = basic_auth['password']
+        auth = (basic_auth_user, basic_auth_password)
+
+        base_url = instance['url']
+        batch_size = instance['default_batch_size']
+        instance_key = {"type": self.INSTANCE_TYPE, "url": base_url}
+        instance_tags = instance.get('tags', [])
+
+        default_timeout = self.init_config.get('default_timeout', 5)
+        timeout = float(instance.get('timeout', default_timeout))
+
+        params = {"service_check_done": False, "batch_size": batch_size, "relation_types": relation_types,
+                  "instance_key": instance_key, "instance_tags": instance_tags, "base_url": base_url,
+                  "auth": auth, "timeout": timeout}
+
+        self._process_and_cache_relation_types(params)
+        self.start_snapshot(instance_key)
+        self._process_components(params)
+        self._process_component_relations(params)
+        self.stop_snapshot(instance_key)
+
+    def _collect_components(self, params):
+        """
+        collect components from ServiceNow CMDB's cmdb_ci table
+        (API Doc- https://developer.servicenow.com/app.do#!/rest_api_doc?v=london&id=r_TableAPI-GET)
+
+        :return: dict, raw response from CMDB
+        """
+
+        base_url = params['base_url']
+        auth = params['auth']
+        timeout = params['timeout']
+        url = base_url + '/api/now/table/cmdb_ci?sysparm_fields=name,sys_id,sys_class_name,sys_created_on'
+
+        return self._get_json(params, url, timeout, auth)
+
+    def _process_components(self, params):
+        """
+        process components fetched from CMDB
+        :return: nothing
+        """
+        instance_tags = params['instance_tags']
+        instance_key = params['instance_key']
+
+        state = self._collect_components(params)
+
+        for component in state['result']:
+            id = component['sys_id']
+            type = {
+                "name": component['sys_class_name']
+            }
+            data = {
+                "name": component['name'].strip(),
+                "tags": instance_tags
+            }
+
+            self.component(instance_key, id, type, data)
+
+    def _collect_relation_types(self, params):
+        """
+        collects relations from CMDB
+        :return: dict, raw response from CMDB
+        """
+
+        base_url = params['base_url']
+        timeout = params['timeout']
+        auth = params['auth']
+        url = base_url + '/api/now/table/cmdb_rel_type?sysparm_fields=sys_id,parent_descriptor'
+
+        return self._get_json(params, url, timeout, auth)
+
+    def _process_and_cache_relation_types(self, params):
```

Comment:
I think caching is not the right word here (the relation types newly retrieved every check run, so we are not really talking about a a cache).

Why are you using the output parameter 'params', I'd much rather see this in a functional style, just return the processed relations from the function